compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeResolver.java

```diff
@@ -0,0 +1,2155 @@
+                    td.pos.lineRange().startLine().offset());
+            if (unknownTypeRefs.add(locationData)) {
+                dlog.error(td.pos, DiagnosticErrorCode.UNKNOWN_TYPE, td.typeName);
+            }
+            return symTable.semanticError;
+        }
+
+        if (moduleLevelDef.getKind() == NodeKind.TYPE_DEFINITION) {
+            BLangTypeDefinition typeDefinition = (BLangTypeDefinition) moduleLevelDef;
+            BType resolvedType = resolveTypeDefinition(symEnv, mod, typeDefinition, depth);
+            if (resolvedType == symTable.semanticError || resolvedType == symTable.noType) {
+                return resolvedType;
+            }
+            symbol = typeDefinition.symbol;
+            td.symbol = symbol;
+            if (symbol.kind == SymbolKind.TYPE_DEF && !Symbols.isFlagOn(symbol.flags, Flags.ANONYMOUS)) {
+                BType referenceType = ((BTypeDefinitionSymbol) symbol).referenceType;
+                referenceType.flags |= symbol.type.flags;
+                referenceType.tsymbol.flags |= symbol.type.flags;
+                return referenceType;
+            }
+            return resolvedType;
+        } else if (moduleLevelDef.getKind() == NodeKind.CONSTANT) {
+            BLangConstant constant = (BLangConstant) moduleLevelDef;
+            resolveConstant(symEnv, mod, constant);
+            return constant.getBType();
+        } else {
+            return extracted(symEnv, (BLangClassDefinition) moduleLevelDef, depth);
+        }
+    }
+
+    private BType resolveTypeDesc(BLangBuiltInRefTypeNode td, SymbolEnv symEnv) {
+        SymbolResolver.AnalyzerData data = new SymbolResolver.AnalyzerData(symEnv);
+        return visitBuiltInTypeNode(td, data, td.typeKind);
+    }
+
+    private BType resolveSingletonType(BLangFiniteTypeNode td, SymbolEnv symEnv) {
+        BTypeSymbol finiteTypeSymbol = Symbols.createTypeSymbol(SymTag.FINITE_TYPE,
+                (Flags.asMask(EnumSet.of(Flag.PUBLIC))), Names.EMPTY, symEnv.enclPkg.symbol.pkgID, null,
+                symEnv.scope.owner, td.pos, BUILTIN);
+
+        // In case we encounter unary expressions in finite type, we will be replacing them with numeric literals.
+         replaceUnaryExprWithNumericLiteral(td);
+
+        BFiniteType finiteType = new BFiniteType(finiteTypeSymbol);
+        for (BLangExpression literal : td.valueSpace) {
+            BType type = blangTypeUpdate(literal);
+            if (type != null && type.tag == TypeTags.SEMANTIC_ERROR) {
+                return type;
+            }
+            if (type != null) {
+                literal.setBType(symTable.getTypeFromTag(type.tag));
+            }
+            finiteType.addValue(literal);
+        }
+        finiteTypeSymbol.type = finiteType;
+        td.setBType(finiteType);
+        return finiteType;
+    }
+
+    private void replaceUnaryExprWithNumericLiteral(BLangFiniteTypeNode finiteTypeNode) {
+        BLangExpression value;
+        NodeKind valueKind;
```

Comment:
Since no usage outside, we can make these local for  for-loop.
Also, finiteTypeNode.valueSpace is repeated can extract to a varaiable